### ISSUE TITLE
[WIP] What I like and what I don't like about Kotlin

### ISSUE THREAD

**Issue by WhiteBlackGoose** [LanguageDev/.NET-language-ideas#20]
I'm not an expert in it at all, but from what I learned

## I dislike

- Poor pattern matching (no way to deconstruct a value)
- Cannot unbox/deconstruct a value after checking its type or nullability, e. g. assume `someFunc()` returns `Int?`, this obviously doesn't work:
```kt
if (someFunc() != null && someFunc().method())
```
If we followed the C# way, it'd be
```kt
if (someFunc() is { } some && some.method())
```
as opposed to
```kt
if (someFunc()?.let { it.method() } != null)
```
- No DUs

## I like

- val and var
- Good syntax for functions (e. g. Int -> (String, Long))
- Enforcement of nullables
- `tailrec`

**Comment by yamin8000** [comment]
> * no way to deconstruct a value

Not 100% true, some values can be deconstructed.

- Pairs
- Data classes

Consider this:
```kotlin
fun getAllCities() : Pair<Result, Error> {
//some heavy work
}

val (cities, error) = getAllCities()
```
The same thing is possible for data classes, however, the order of values is the same as the definition. This is also usable for lambdas, observe:
```kotlin
myListItem.setOnClickListener { (index,item) ->
//handle item
}

```
Anyway as the philosophy of the Fresh Language is to make C# fresh again as Kotlin did to Java I guess as similar both Java and C# are many features of Kotlin can be added to it.
Extension methods are a must feature.

**Comment by WhiteBlackGoose** [comment]
> Consider this:

This is not pattern matching. Deconstructing in pattern matching would be
```cs
a switch
{
    (5, var value) => value * 2
}
```

> Extension methods are a must feature

Well, they exist in C# too, but yeah, it'd be nice to have them as in Kotlin. If we don't have an issue for it, feel free to open one

**Comment by yamin8000** [comment]
> > Consider this:
> 
> This is not pattern matching. Deconstructing in pattern matching would be
> 
> ```cs
> a switch
> {
>     (5, var value) => value * 2
> }
> ```
> 
> > Extension methods are a must feature
> 
> Well, they exist in C# too, but yeah, it'd be nice to have them as in Kotlin. If we don't have an issue for it, feel free to open one

I agree, Kotlin has some sort of deconstructing but fails in pattern matching.
